---
title: "DEHP_CELfiles"
author: "Shrija Srinivasan"
date: "5/27/2020"
output: html_document
---

```{r setup installation of libraries, include=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.12")

if (!require("BiocManager"))
    install.packages("BiocManager")
BiocManager::install("maEndToEnd", version = "devel")

## Downloading the necessary packages to make maEndToEnd work, installing the devtools and remote packages
install.packages("devtools")
devtools::install_github("r-lib/remotes")
packageVersion("remotes") # has to be 1.1.1.9000 or later

#Installing remotes from github
remotes::install_github("b-klaus/maEndToEnd", ref="master")


if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ArrayExpress")

# Installing WCGNA for correlation
BiocManager::install("WCGNA")
BiocManager::install("GenomicRanges")
BiocManager::install("SummarizedExperiment")
BiocManager::install("DESeq2")
BiocManager::install("tidyverse")

# The following initializes usage of Bioc devel
BiocManager::install(version='devel')
BiocManager::install("AnnotationDbi")

# The following initializes usage of Bioc devel
BiocManager::install("ragene11sttranscriptcluster.db")
BiocManager::install("ggplot")
BiocManager::install("apeglm")
BiocManager::install("GEOquery")
BiocManager::install("edgeR")
BiocManager::install.packages("stringr")

## Calling out the library functions
# General Bioconductor packages
library(Biobase)
library(oligoClasses)

# Annotation and data import packages
library(devtools)
library(remotes)
suppressPackageStartupMessages({library("maEndToEnd")})
library(ArrayExpress)
library("AnnotationDbi")
library("org.Rn.eg.db")
columns(org.Rn.eg.db)
library(pd.hugene.1.0.st.v1)
library(ragene11sttranscriptcluster.db)
library(stringr)

# Quality control and pre-processing packages
library(limma)
library(umap)
library(topGO)
library(ReactomePA)
library(clusterProfiler)
library(tidyverse)
library(edgeR)
library(DESeq2)
library(apeglm)

# Plotting and color options packages
library(ggplot2)
library(geneplotter)
library(RColorBrewer)
library(pheatmap)


# Formatting/documentation packages
library(dplyr)
library(tidyr)

# Helpers
library(stringr)
library(matrixStats)
library(genefilter)
library(statmod)

# Downloading raw data from ArrayExpress
raw_data_dir <- tempdir()
if (!dir.exists(raw_data_dir)) {
    dir.create(raw_data_dir)
}


# File with BPA: fetching and loading the annotation file
anno_AE <- getAE("E-MTAB-4458", path = raw_data_dir, type = "raw")


# Importing annotation data and microarray expression dataset (incase fetching doesn't work)
# Providing the required file path for extracting the data
sdrf_location <- file.path(raw_data_dir, "E-MTAB-4458.sdrf.txt")
sdrf_location
sdrf <- read.delim(sdrf_location, check.names = TRUE, as.is = TRUE)
rownames(sdrf) <- sdrf$Array.Data.File
sdrf<- AnnotatedDataFrame(sdrf)

raw_data <- oligo::read.celfiles(filenames = file.path(raw_data_dir, sdrf$Array.Data.File),verbose = FALSE, phenoData = sdrf)
stopifnot(validObject(raw_data))


# To see how the raw data looks like (a sample):
head(Biobase::pData(raw_data))
pData(raw_data)
Biobase::pData(raw_data) <- Biobase::pData(raw_data)[,c("Source.Name","Characteristics.time.","Factor.Value.compound.",
"Factor.Value.dose.")]


#rEMA
# Quality control of the raw data
countraw <- Biobase::exprs(raw_data)
metadata <- Biobase::pData(raw_data)
summary(Biobase::exprs(raw_data))


# Check if the data is log2 transformed or not (Checking individual datasets):
oligo::boxplot(countraw, target="core", main ="Boxplot of log2-intensities for the raw data")


## Remove genes with low counts
keepgenes <- rowSums(edgeR::cpm(countraw)>1) >= 2
countfiltered <- countraw[keepgenes,]


## log2 scale data pre-processing
exp_raw <- log2(Biobase::exprs(raw_data))
PCA_raw <- prcomp(t(exp_raw), scale.=FALSE)
exp_raw.umap <- umap(t(exp_raw))
head(exp_raw.umap$layout,3)


#Considering the 10% most variable 
percentVar <- round(100*PCA_raw$sdev^2/sum(PCA_raw$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])


# Plotting the PCA (log-transformed data)
dataGG <- data.frame(PC1 = PCA_raw$x[,1], PC2 = PCA_raw$x[,2],
                    Compound = pData(raw_data)$Factor.Value.compound.,
                    Time = pData(raw_data)$Characteristics.time.,
                    Dose = pData(raw_data)$Factor.Value.dose.)
ggplot(dataGG, aes(PC1, PC2)) + geom_point(aes(shape =Compound,colour = Time))+ ggtitle("PCA plot of the log-transformed raw expression data (BPA)")+ xlab(paste0("PC1"))+ ylab(paste0("PC2"))+ theme(plot.title = element_text(hjust = 0.5))


# Since we are dealing with oligo method, there are decent in size and can be deconvoluted (to original high quality) for better summarisation, background correction and the RMA
eset <- oligo::rma(raw_data, target = "core")


##    Plotting RLE

rowMedians_data <- Biobase::rowMedians(as.matrix(Biobase::exprs(eset)))
RLE_BPAdata <- sweep(Biobase::exprs(eset),1,rowMedians_data)
RLE_BPAdata <- as.data.frame(RLE_BPAdata)
RLE_gathereddata <- tidyr::gather(RLE_BPAdata, BPAcontrol_array, log2_expression_deviation)

#Plotting RLE graph

ggplot2:: ggplot(RLE_gathereddata, aes(BPAcontrol_array, log2_expression_deviation))+geom_boxplot(outlier.shape = NA) + ylim(c(-2, 2)) + theme(axis.text.x = element_text(colour = "aquamarine4", angle = 60, size = 6.5, hjust = 1 , face = "bold"))


##  Quality assessment of the calibrated data

exp_eset <- Biobase::exprs(eset)
pData(eset)
PCA<- prcomp(t(exp_eset),scale. = FALSE)
percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)
sd_ratio <- sqrt(percentVar[2]/percentVar[1])
dataGG <- data.frame(PC1 = PCA$x[,1], PC2= PCA$x[,2],
                     Compound = pData(eset)$Factor.Value.compound.,
                     Time = pData(eset)$Characteristics.time.,
                     Dose= pData(eset)$Factor.Value.dose.)
ggplot(dataGG, aes(PC1,PC2)) + geom_point(aes(shape =Compound, colour =Time)) + ggtitle("PCA plot of the calibrated, summarised data") + xlab(paste0("PC1, VarExp:", percentVar[1],"%")) + ylab(paste0("PC2, VarExp:", percentVar[2],"%")) +theme(plot.title = element_text(hjust = 0.5))

## Heatmap Clustering analysis

compoundnames <- ifelse(str_detect(pData(eset)$Factor.Value.compound.,"bisphenol A"),"bisphenol A","control")

annotation_for_heatmap <- data.frame(Time = pData(eset)$Characteristics.time.,Compound = compoundnames)
row.names(annotation_for_heatmap) <- row.names(pData(eset))

dists <- as.matrix(dist(t(exp_eset),method = "manhattan"))
rownames(dists) <- row.names(pData(eset))
hmcol <- rev(colorRampPalette(RColorBrewer::brewer.pal(9, "YlOrRd"))(255))
colnames(dists) <- NULL
diag(dists) <- NA


ann_colors <- list(
  Compound = c("bisphenol A" = "chartreuse4", "control" = "burlywood3"),
  Time = c("1" = "blue4", "3" = "cadetblue2", "7" ="white"))
colnames(dists) <- colnames(dists)


pheatmap(dists, col = (hmcol),
         annotation_colors = ann_colors,
         legend = TRUE, 
         treeheight_row = 0,
         legend_breaks = c(min(dists, na.rm = TRUE), 
                         max(dists, na.rm = TRUE)), 
         legend_labels = (c("small distance", "large distance")),
         main = "Clustering heatmap for the calibrated samples")


dists <- as.matrix(dist(t(exp_eset), method = "manhattan"))
rownames(dists) <- row.names(pData(eset))
hmcol <- rev(colorRampPalette(RColorBrewer::brewer.pal(9, "YlOrRd"))(255))
colnames(dists) <- NULL
diag(dists) <- NA


ann_colors <- list(
  Compound = c("bisphenol A" = "chartreuse4", "control" = "burlywood3"),
  Time = c("1" = "blue4", "3" = "cadetblue2", "7" ="white"))
colnames(dists) <- colnames(dists)


pheatmap(dists, col = (hmcol), 
         annotation_row = annotation_for_heatmap,
         annotation_colors = ann_colors,
         legend = TRUE, 
         treeheight_row = 0,
         legend_breaks = c(min(dists, na.rm = TRUE), 
                         max(dists, na.rm = TRUE)), 
         legend_labels = (c("small distance", "large distance")),
         main = "Clustering heatmap for the calibrated samples")

# Filtering based on intensity (log2[intensity <=5])
# Linear model to check which genes are differentially expressed between control and the celllines
medianint <- rowMedians(Biobase::exprs(eset))
hist_res <- hist(medianint, 100, col="cornsilk" , freq = FALSE, main= "Histogram of the median intensities", border = "antiquewhite4", xlab =" Median intensities")

man_threshold <- 5
hist_res <- hist(medianint, 100, col="cornsilk", freq= FALSE, main= "Histogram of the median intensities", border ="antiquewhite4", xlab ="Median intensities")
abline(v= man_threshold, col="coral4",lwd=2)


number_of_samples <- table(paste0(pData(eset)$Factor.Value.compound.,"_", pData(eset)$Characteristics.time.))
number_of_samples

samples_cutoff <- min(number_of_samples)
samples_cutoff

idx_man_threshold <- apply(Biobase::exprs(eset), 1,
                           function(x){
                          sum(x > man_threshold) >= samples_cutoff})

table(idx_man_threshold)
idx_man_threshold

manfiltered <- subset(eset, idx_man_threshold)
manfiltered

## Annotation of the transcript clusters
# Annotate the datasets

library(ragene11sttranscriptcluster.db)
#head(ls("package::ragene10sttranscriptcluster.db"))
annotation_eset <- AnnotationDbi::select(ragene11sttranscriptcluster.db,
                                  keys = (featureNames(manfiltered)),
                                  columns = c("SYMBOL", "GENENAME"),
                                  keytype = "PROBEID")

# Drop NA values
annotation_eset <- subset(annotation_eset, !is.na(SYMBOL))
annotation_eset

# Remove multiple mappings/ removing redundancies
annotation_grouped <- group_by(annotation_eset, PROBEID)
annotation_grouped
annotation_summarised <- dplyr::summarize(annotation_grouped, no_of_matches = n_distinct(SYMBOL))
head(annotation_summarised)


annotation_filtered <- filter(annotation_summarised, no_of_matches == 1)
head(annotation_filtered)
table(annotation_filtered$no_of_matches)
probe_stats <- annotation_filtered
probe_stats
dim(probe_stats)

# We have to include these unwanted ProbeID from feature data too
ids_to_exclude <- (featureNames(manfiltered) %in% probe_stats$PROBEID)
ids_to_exclude
table(ids_to_exclude)
finallist <- subset(manfiltered, ids_to_exclude)
validObject(finallist)

## g <- avereps(fData(finallist),fData(finallist)$PROBEID)
#head(g)

fData(finallist)$PROBEID <- rownames(fData(finallist))

#match(fData(finallist)$PROBEID, annotation_eset$PROBEID, nomatch = 0)
#fData(finallist)$PROBEID[1] %in% annotation_eset$PROBEID

fData(finallist) <- left_join( fData(finallist), annotation_eset)
fData(finallist)
Biobase::pData(finallist)$Source.Name

# Restoring the rownames after merging
rownames(fData(finallist)) <- fData(finallist)$PROBEID
validObject(finallist)

## Remove probe replicates
data_summarised <- limma::avereps(finallist)

## Linear Model of the Data (LIMMA)
# Since we are working with time-series data, my plan is to design an all-to-all comparison matrix (no intercept). Tale all groups separately and collapse the sourcename.

sourcename <- as.character(Biobase::pData(finallist)$Source.Name)
sampletime <- as.character(Biobase:: pData(finallist)$Characteristics.time.)
compound <- str_replace_all(Biobase::pData(finallist)$Factor.Value.compound.," ","_")
compound <- ifelse(compound == "bisphenol_A","BPA","control")


# Now designing the Limma model
# Define BPA groups

design_BPA <- model.matrix(~0+sampletime[compound == "BPA"])

# By making some space for control groups
design_BPA <- rbind(design_BPA,matrix(rep(0,24), nrow=6, ncol=3))
design_BPA<-cbind(design_BPA,matrix(rep(0,30), nrow=15, ncol=2))

# Define control groups
design_control <- model.matrix(~0+sampletime[compound == "control"])
design_BPA[10:15,4:5]<-design_control

# Name columns
colnames(design_BPA) <- c("BPA1", "BPA3","BPA7","CTL1","CTL7")

fit <- lmFit(data_summarised, design_BPA )
contrast.matrix <- makeContrasts(CTL1-BPA1, CTL1-BPA3, CTL1-BPA7, CTL7-BPA1, CTL7-BPA3, CTL7-BPA7, levels=design_BPA)

fit2 <- contrasts.fit(fit, contrast.matrix)
fit2 <- eBayes(fit2)

#top genes differentially expressed in CTL1 versus BPA1 (the comparison depends on the coef)
topTable(fit2, coef=1, adjust="BH")


#The outcome of each hypothesis test can be assigned using
results <- decideTests(fit2)

#A Venn diagram showing numbers of genes significant in each comparison with CTL1
vennDiagram(results[,1:3])

table_fit2 <- topTable(fit2, number = Inf)
head(table_fit2)

# Plotting a histogram accordingly
hist(table_fit2$P.Value, col= brewer.pal(3, name= "Set2")[1], main = "BPA vs Control", xlab ="p-values")
tail(subset(table_fit2,P.Value < 0.001))

nrow(subset(table_fit2, P.Value < 0.001))

# Visualisation of DE Analysis Results- Volcano Plot
volcanonames <- ifelse(abs(fit2$coefficients)>=1, fData(finallist)$SYMBOL, NA) 
volcanoplot(fit2, coef = 1L ,style = "p-value", highlight = 15, xlab = "log2Fold Change", ylab = NULL, pch =16, cex = 0.35)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
